### PR TITLE
Remove Boost.Program_options from the programs (fix #33)

### LINKED ISSUE #33: Remove boost::program options
Boost::program_options is incompatible with old gcc, clang and builds fail

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=797917

### PR DESCRIPTION
**Pull request by krivenko**
A copy of https://github.com/Taywee/args/blob/master/args.hxx has been downloaded and put into `prog/`.

`args.hxx`'s API is a bit different from Boost's, so I had to redesign the `quantum_model` class and its derivatives.

### CODE REVIEW

**Comment by krivenko** [comment]
All four jobs in the Travis CI matrix pass now!

**Review by aeantipov — APPROVED** [review]

File: prog/anderson_model.h [hunk 1]
```diff
@@ -112,12 +82,19 @@ class anderson_model: public quantum_model {
 
       mpi_cout << "Terms with 4 operators" << std::endl;
       Lat.printTerms(4);
-    };
+    }
   }
+
+  struct {
```
Comment (aeantipov) [inline prog/anderson_model.h:88]: is this a new recommended way of encapsulating parameters into structs and fetching them with std::get?

**Comment by aeantipov** [comment]
I like this `args.hxx` thing and I like the cleaner arrangement of parameters.

**Review by aeantipov — APPROVED** [review]

File: prog/anderson_model.h (hunk 1 above)
Comment (krivenko) [reply prog/anderson_model.h:88]: This structure is essentially a namespace that isolates args option objects from eponymous data members (c.f. `anderson_model::levels` and `anderson_model::args_options.levels`).

**Pull request merged**